### PR TITLE
feat!: throttle with retries provider changes

### PR DESCRIPTION
**Pull request by smarshall-spitzbart**
## Description

- Closes: https://github.com/cosmos/interchain-security/issues/1102, part of  https://github.com/cosmos/interchain-security/issues/713
- Updates/adds UTs and integration tests as needed to close 1102
- Implicitly reverts https://github.com/cosmos/interchain-security/pull/1004
- Depreciates provider's `QueryThrottledConsumerPacketData` query, an equivalent query will be added to consumer, tracked as https://github.com/cosmos/interchain-security/issues/1233
- Updates throttling related e2e tests to pass. Note retry logic is not fully e2e tested in this PR, and will be fully addressed, tracked by https://github.com/cosmos/interchain-security/issues/1103 
- Migration is NOT addressed in this PR, see #1032 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [x] Added `!` to the type prefix if state-machine breaking change (i.e., requires coordinated upgrade)
* [ ] Confirmed this PR does not introduce changes requiring state migrations, OR migration code has been added to consumer and/or provider modules
* [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
* [x] Provided a link to the relevant issue or specification
* [x] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules)
* [x] Included the necessary unit and integration [tests](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#testing)
* [x] Added a changelog entry to `CHANGELOG.md`
* [x] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [x] Updated the relevant documentation or specification
* [x] Reviewed "Files changed" and left comments if necessary 
* [ ] Confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] confirmed this PR does not introduce changes requiring state migrations, OR confirmed migration code has been added to consumer and/or provider modules
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

### CODE REVIEW

File: tests/integration/stop_consumer.go [hunk 1]
```diff
@@ -80,33 +79,6 @@ func (s *CCVTestSuite) TestStopConsumerChain() {
 				return nil
 			},
 		},
-		{
-			func(suite *CCVTestSuite) error {
```
Comment (smarshall-spitzbart) [inline tests/integration/stop_consumer.go:84]: Any test or test case that was removed, is a test that is only relevant to provider queuing logic which is removed in this PR

File: tests/integration/throttle_retry.go [hunk 2]
```diff
@@ -1,32 +1,25 @@
...
-// TestSlashRetries tests the throttling v2 retry logic. Without provider changes,
-// the consumer will queue up a slash packet, the provider will return a v1 result,
-// and the consumer will never need to retry.
-//
-// Once provider changes are made (slash packet queuing is removed), the consumer may retry packets
-// via new result acks from the provider.
-//
-// TODO: This test will need updating once provider changes are made.
+// TestSlashRetries tests the throttling v2 retry logic at an integration level.
 func (s *CCVTestSuite) TestSlashRetries() {
```
Comment (smarshall-spitzbart) [inline tests/integration/throttle_retry.go:14]: This test has been adjusted to validate the full throttling v2 logic (provider/consumer both having v2 throttling), previously it just tested that a consumer running the v2 throttling protocol is compatible with a provider running the v1 throttling protocol

File: x/ccv/provider/keeper/relay.go [hunk 3]
```diff
@@ -270,18 +238,10 @@ func (k Keeper) QueueVSCPackets(ctx sdk.Context) {
...
-	blockHeight := uint64(ctx.BlockHeight()) + 1
-	valUpdateID := k.GetValidatorSetUpdateId(ctx)
-	k.SetValsetUpdateBlockHeight(ctx, valUpdateID, blockHeight)
-	k.Logger(ctx).Debug("vscID was mapped to block height", "vscID", valUpdateID, "height", blockHeight)
-
-	// Replenish slash meter if necessary, BEFORE executing slash packet throttling logic.
-	// This ensures the meter value is replenished, and not greater than the allowance (max value)
-	// for the block, before the throttling logic is executed.
+// BeginBlockCIS contains the BeginBlock logic needed for the Consumer Initiated Slashing sub-protocol.
+func (k Keeper) BeginBlockCIS(ctx sdk.Context) {
```
Comment (smarshall-spitzbart) [inline x/ccv/provider/keeper/relay.go:242]: Slash meter replenishment now happens on beginblock, since we want to replenish the meter before slash packets are handled DURING the block. Whereas previously all slash packets were handled in endblocker after the replenishment call

File: x/ccv/provider/keeper/relay.go [hunk 4]
```diff
@@ -358,28 +311,36 @@ func (k Keeper) OnRecvSlashPacket(ctx sdk.Context, packet channeltypes.Packet, d
...
-		ctx.BlockTime(),   // recv time
-		chainID,           // consumer chain id that sent the packet
-		packet.Sequence,   // IBC sequence number of the packet
-		providerConsAddr)) // Provider consensus address of val to be slashed
-
-	// Queue slash packet data in the same (consumer chain specific) queue as vsc matured packet data,
-	// to enforce order of handling between the two packet data types.
-	if err := k.QueueThrottledSlashPacketData(ctx, chainID, packet.Sequence, data); err != nil {
-		return ccv.NewErrorAcknowledgementWithLog(ctx, fmt.Errorf("failed to queue slash packet data: %s", err.Error()))
+	meter := k.GetSlashMeter(ctx)
```
Comment (smarshall-spitzbart) [inline x/ccv/provider/keeper/relay.go:314]: Meat of provider changes is here. If slash meter is negative, bounce packet, otherwise decrement meter and handle packet

**Comment by smarshall-spitzbart** [comment]
This PR should be based onto `feat/throttle-with-retries-complete`, but I'm not able to update that base branch to reflect merges to main.. This PR is ready to review regardless

File: tests/e2e/actions.go [hunk 5]
```diff
@@ -1935,50 +1935,41 @@ func (tr TestRun) assignConsumerPubKey(action assignConsumerPubKeyAction, verbos
...
+// slashMeterReplenishmentAction polls the slash meter on provider until value is achieved
+type slashMeterReplenishmentAction struct {
+	targetValue int64
 	// panic if timeout is exceeded
 	timeout time.Duration
 }
 
-func (tr TestRun) waitForSlashThrottleDequeue(
-	action slashThrottleDequeue,
+func (tr TestRun) waitForSlashMeterReplenishment(
```
Comment (smarshall-spitzbart) [inline tests/e2e/actions.go:1945]: The slash packet queue size of the provider is no longer relevant, but the slash meter value is relevant. I've replaced this action to wait on slash meter being replenished

File: tests/e2e/steps_downtime.go [hunk 6]
```diff
@@ -364,27 +358,24 @@ func stepsThrottledDowntime(consumerName string) []Step {
...
 					ValPowers: &map[validatorID]uint{
 						validatorID("alice"): 511,
-						validatorID("bob"):   0,
+						validatorID("bob"):   0, // VSC packet applying bob jailing is also relayed and recv by consumer
 						validatorID("carol"): 500,
 					},
 				},
 			},
 		},
+		// TODO(Shawn): Improve this test to have the consumer retry it's downtime slash, and to assert queue size on consumer.
```
Comment (smarshall-spitzbart) [inline tests/e2e/steps_downtime.go:373]: Note this TODO, the e2e test ends without retrying the slash packet. This will be addressed in a separate PR

**Review by mpoke — COMMENTED** [review]
Nice work @smarshall-spitzbart. I approve the logic, didn't look into the tests though. 

Nit: do you know why the labels indicate the PR contains changes to the consumer?

**Comment by smarshall-spitzbart** [comment]
> do you know why the labels indicate the PR contains changes to the consumer?

Not sure, I don't see any changes to the consumer, it might have to do with merging main on this branch and how many times I had to change the base (feat) branch to reflect changes in main

File: CHANGELOG.md [hunk 7]
```diff
@@ -4,6 +4,7 @@
 
 Add an entry to the unreleased section whenever merging a PR to main that is not targeted at a specific release. These entries will eventually be included in a release.
 
+* (feat!) throttle with retries provider changes [#1230](https://github.com/cosmos/interchain-security/pull/1230)
```
[the remark this answers is not in the document]
Comment (smarshall-spitzbart) [reply CHANGELOG.md:7]: Sure I'll need https://github.com/cosmos/interchain-security/pull/1267 merged to do so

**Review by jtremback — APPROVED** [review]
LGTM but I didn't really review the tests. Great work!

File: tests/e2e/actions.go [hunk 8]
```diff
@@ -1937,50 +1937,41 @@ func (tr TestRun) assignConsumerPubKey(action assignConsumerPubKeyAction, verbos
...
+		if slashMeter >= action.targetValue {
 			break
 		}
 
 		if time.Now().After(timeout) {
-			panic(fmt.Sprintf("\n\n\nwaitForSlashThrottleDequeuemethod has timed out after: %s\n\n", action.timeout))
+			panic(fmt.Sprintf("\n\nwaitForSlashMeterReplenishment has timed out after: %s\n\n", action.timeout))
 		}
 
 		time.Sleep(500 * time.Millisecond)
```
Comment (p-offtermatt) [inline tests/e2e/actions.go:1973]: Can we replace this with tr.waitBlocks/tr.waitTime?

Comment (p-offtermatt) [reply tests/e2e/actions.go:1973]: I think it should be waitTime from my understanding of the slashMeter (it fills by timestamp, not by blockcount?)

File: tests/e2e/steps_downtime.go [hunk 9]
```diff
@@ -343,10 +338,9 @@ func stepsThrottledDowntime(consumerName string) []Step {
 					},
 				},
 				chainID(consumerName): ChainState{
-					// VSC packet applying jailing is not yet relayed to consumer
 					ValPowers: &map[validatorID]uint{
 						validatorID("alice"): 511,
-						validatorID("bob"):   500,
+						validatorID("bob"):   500, // VSC packet applying bob jailing is not yet relayed to consumer
```
Comment (p-offtermatt) [inline tests/e2e/steps_downtime.go:343]:
```suggestion
						validatorID("bob"):   500, // VSC packet jailing bob is not yet relayed to consumer
```

File: tests/e2e/state.go [hunk 10]
```diff
@@ -667,9 +654,10 @@ func (tr TestRun) getProviderAddressFromConsumer(consumerChain chainID, validato
 	return addr
 }
 
-func (tr TestRun) getGlobalSlashQueueSize() uint {
+func (tr TestRun) getSlashMeter() int64 {
```
Comment (p-offtermatt) [inline tests/e2e/state.go:657]: Should this take the chain as an argument instead of always using provi?

File: tests/e2e/steps_downtime.go (hunk 6 above)
Comment (p-offtermatt) [reply tests/e2e/steps_downtime.go:373]: Just to confirm, we are removing the Queue size from the chain state, despite this TODO, right? It reads to me as if that would reintroduce it, but maybe I'm missing context

File: tests/e2e/steps_downtime.go [hunk 11]
```diff
@@ -364,42 +358,35 @@ func stepsThrottledDowntime(consumerName string) []Step {
...
-				nextQueueSize:    0,
+			action: slashMeterReplenishmentAction{
+				targetValue: 0, // We just want slash meter to be non-negative
+
 				// Slash meter replenish fraction is set to 10%, replenish period is 20 seconds, see config.go
 				// Meter is initially at 10%, decremented to -23% from bob being jailed. It'll then take three replenishments
-				// for meter to become positive again. 3*20 = 60 seconds + buffer = 80 seconds
-				timeout: 80 * time.Second,
+				// for meter to become positive again. 3*20 = 60 seconds + buffer = 100 seconds
+				timeout: 100 * time.Second,
```
Comment (p-offtermatt) [inline tests/e2e/steps_downtime.go:382]: Just curious, was there a reason the buffer needs to be increased?

**Review by p-offtermatt — COMMENTED** [review]
First part of review, more reviewing to come

**Review by p-offtermatt — APPROVED** [review]
I added some more comments. Long list of removed code, nothing I was looked seriously problematic, so I'll approve. I mostly looked at the tests, but it was hard for me to tell whether something that should be tested is untested

File: tests/e2e/actions.go [hunk 12]
```diff
@@ -1937,50 +1937,41 @@ func (tr TestRun) assignConsumerPubKey(action assignConsumerPubKeyAction, verbos
...
 	verbose bool,
 ) {
 	timeout := time.Now().Add(action.timeout)
-	initialGlobalQueueSize := int(tr.getGlobalSlashQueueSize())
+	initialSlashMeter := tr.getSlashMeter()
 
-	if initialGlobalQueueSize != action.currentQueueSize {
-		panic(fmt.Sprintf("wrong initial queue size: %d - expected global queue: %d\n", initialGlobalQueueSize, action.currentQueueSize))
+	if initialSlashMeter >= 0 {
+		panic(fmt.Sprintf("No need to wait for slash meter replenishment, current value: %d", initialSlashMeter))
```
Comment (p-offtermatt) [inline tests/e2e/actions.go:1955]: Is this a bad idea to panic on? It makes the tests flakey, because if they happen to run slowly, the slash meter might be already-full, right?

File: tests/e2e/state.go (hunk 10 above)
Comment (smarshall-spitzbart) [reply tests/e2e/state.go:657]: Imo no because the slash meter is only relevant to the provider. If we ever plan to have multiple providers in these tests then that's a different story, but I don't see that happening

File: tests/e2e/steps_downtime.go (hunk 11 above)
Comment (smarshall-spitzbart) [reply tests/e2e/steps_downtime.go:383]: I think this is an artifact of https://github.com/cosmos/interchain-security/issues/1304 and trying to get this to work with cometmock

File: tests/e2e/actions.go (hunk 8 above)
Comment (smarshall-spitzbart) [reply tests/e2e/actions.go]: Nice catch https://github.com/cosmos/interchain-security/commit/53c02ea95260788b9f42a881cb3939e5c253b919

File: tests/e2e/steps_downtime.go (hunk 6 above)
Comment (smarshall-spitzbart) [reply tests/e2e/steps_downtime.go:374]: Good question, we are removing the queues _from the provider_ in this PR. The queues exist on the consumers as of https://github.com/cosmos/interchain-security/pull/1024, so incoming e2e test will assert queue size on consumer

File: x/ccv/provider/keeper/grpc_query.go [hunk 13]
```diff
@@ -150,24 +149,6 @@ func (k Keeper) QueryThrottleState(goCtx context.Context, req *types.QueryThrott
 	candidate := k.GetSlashMeterReplenishTimeCandidate(ctx) // always UTC
 	packets := []*types.ThrottledSlashPacket{}
```
[the remark this answers is not in the document]
Comment (smarshall-spitzbart) [reply x/ccv/provider/keeper/grpc_query.go]: Yea my original intention here was to avoid breaking the query API. But you make a good point that this PR is already "query breaking", might as well make things clean -> https://github.com/cosmos/interchain-security/commit/1ceddcb356493162875d448b5fef2dfb07859a10

[the remark this answers is not in the document]
Comment (smarshall-spitzbart) [reply x/ccv/provider/keeper/grpc_query.go]: To be consistent I should also remove this deprecated provider query too -> https://github.com/cosmos/interchain-security/commit/e228953c82039b3a7ac20343899af89cabc2afad

File: x/ccv/provider/keeper/grpc_query.go [hunk 14]
```diff
@@ -191,55 +172,14 @@ func (k Keeper) QueryThrottledConsumerPacketData(goCtx context.Context, req *typ
...
-			})
-		default:
-			k.Logger(ctx).Error(fmt.Sprintf("unexpected packet data type: %T", data))
-		}
-	}
 
 	return &types.QueryThrottledConsumerPacketDataResponse{
 		ChainId:             req.ChainId,
-		Size_:               k.GetThrottledPacketDataSize(ctx, req.ChainId),
+		Size_:               0,
```
[the remark this answers is not in the document]
Comment (smarshall-spitzbart) [reply x/ccv/provider/keeper/grpc_query.go]: I removed this entire query

File: x/ccv/provider/keeper/proposal_test.go [hunk 15]
```diff
@@ -525,28 +524,6 @@ func TestStopConsumerChain(t *testing.T) {
 			},
 			expErr: true,
 		},
-		{
-			description: "valid stop of consumer chain, throttle related queues are cleaned",
```
[the remark this answers is not in the document]
Comment (smarshall-spitzbart) [reply x/ccv/provider/keeper/proposal_test.go:529]: I beleive the entire intention of this test case was to confirm the provider queues  are cleared. Since those queues no longer exist it seems neccessary to remove the whole test case

File: tests/e2e/actions.go (hunk 12 above)
Comment (smarshall-spitzbart) [reply tests/e2e/actions.go:1955]: Hmm I don't see a good way to decouple the logic of the slash meter away from time. Any test for throttling must have a pretty good and mostly deterministic control over chain time
